fabric-client-tags-api-v1/src/client/java/net/fabricmc/fabric/api/tag/client/v1/ClientTags.java

```diff
@@ -60,13 +62,84 @@ public static Set<Identifier> getOrCreateLocalTag(TagKey<?> tagKey) {
 		Set<Identifier> ids = LOCAL_TAG_CACHE.get(tagKey);
 
 		if (ids == null) {
-			ids = ClientTagsLoader.loadTag(tagKey);
+			ClientTagsLoader.LoadedTag loadedTag = ClientTagsLoader.loadTag(tagKey);
+			ids = loadedTag.completeIds();
 			LOCAL_TAG_CACHE.put(tagKey, ids);
+			LOCAL_TAG_HIERARCHY.put(tagKey, loadedTag.childHolder());
 		}
 
 		return ids;
 	}
 
+	//todo should this be exposed and getOrCreateLocalTag deprecated?
+	private static ClientTagsLoader.ChildHolder getOrCreatePartiallySyncedTag(TagKey<?> tagKey) {
+		ClientTagsLoader.ChildHolder childHolder = LOCAL_TAG_HIERARCHY.get(tagKey);
+
+		if (childHolder == null) {
+			ClientTagsLoader.LoadedTag loadedTag = ClientTagsLoader.loadTag(tagKey);
+			childHolder = loadedTag.childHolder();
+			LOCAL_TAG_CACHE.put(tagKey, loadedTag.completeIds());
+			LOCAL_TAG_HIERARCHY.put(tagKey, childHolder);
+		}
+
+		return childHolder;
+	}
+
+	//todo this should probably be the default behavior of *withLocalFallback, with new way to do the old behavior
+	// then again, arguably the javadoc describes the old behavior
+	public static <T> boolean isInPartiallySyncedTag(TagKey<T> tagKey, RegistryEntry<T> registryEntry) {
+		Objects.requireNonNull(tagKey);
+		Objects.requireNonNull(registryEntry);
+
+		// Check if the tag exists in the dynamic registry first
+		Optional<? extends Registry<T>> maybeRegistry = getRegistry(tagKey);
+
+		if (maybeRegistry.isPresent()) {
+			if (maybeRegistry.get().getEntryList(tagKey).isPresent()) {
+				return registryEntry.isIn(tagKey);
+			}
+		}
+
+		if (registryEntry.getKey().isPresent()) {
+			ClientTagsLoader.ChildHolder wt = getOrCreatePartiallySyncedTag(tagKey);
+			boolean isIn = wt.immediateChildIds().contains(registryEntry.getKey().get().getValue());
+			Iterator<TagKey<?>> it = wt.immediateChildTags().iterator();
+
+			while (!isIn && it.hasNext()) {
+				isIn = isInPartiallySyncedTag((TagKey<T>) it.next(), registryEntry);
+			}
+
+			return isIn;
+		}
+
+		return false;
+	}
+
+	public static <T> boolean isInPartiallySyncedTag(TagKey<T> tagKey, T entry) {
```

Comment:
Can't we put the logic to retrieve `RegistryEntry<T>` from `TagKey` and `T` in a shared private helper to avoid repeating it?